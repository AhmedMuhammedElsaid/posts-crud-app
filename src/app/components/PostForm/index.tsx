"use client";
import React, { useEffect, useState } from "react";
import classes from "./styles.module.css";
import { toast } from "react-toastify";
import axios from "axios";
import { IPropsEdit } from "@/app/types";

const PostForm: React.FC<IPropsEdit> = ({ id }) => {
  const [title, setTitle] = useState("");
  const [body, setBody] = useState("");
  const [isLoading, setIsLoading] = useState(false);

  useEffect(() => {
    if (!id) return;
    // Fetch the post and set the title and body in case of edit mode
    (async function fetchPost() {
      const postData = await axios.get(
        `https://jsonplaceholder.typicode.com/posts/${id}`
      );
      if (postData?.status === 200) {
        setTitle(postData.data.title);
        setBody(postData.data.body);
      }
    })();
  }, [id]);

  const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    // validate inputs before submitting
    if (handleValidateInputs()) return;

    setIsLoading(true);

    const postData = { title, body, userId: 1 };

    const response = await axios(
      `https://jsonplaceholder.typicode.com/posts${id ? `/${id}` : ""}`,
      {
        method: id ? "PUT" : "POST",
        data: id ? { ...postData, id } : postData,
      }
    );

    if ([200, 201].includes(response.status)) {
      toast.success(
        `Post ${id ? `Number ${id} Edited` : "Created"} Successfully`
      );
      setIsLoading(false);
      if (id) return;
      setBody("");
      setTitle("");
    } else {
      toast.error("Something went wrong. please try again");
      setIsLoading(false);
    }
  };
  const handleValidateInputs = () => {
    if (!title.length) {
      toast.error("Please enter the post title");
      return true;
    }
    if (!body.length) {
      toast.error("Please enter the post body");
      return true;
    }
    return false;
  };

  return (
    <form className={classes.form} onSubmit={handleSubmit}>
      <label className={classes.label} htmlFor="body">
        Title
      </label>
      <input
        className={classes.input}
        type="text"
        value={title}
        onChange={(e) => setTitle(e.target.value)}
        placeholder="write post title..."
        data-testid="title-input"
      />
      <label className={classes.label} htmlFor="body">
        Body
      </label>
      <textarea
        className={classes.textarea}
        value={body}
        onChange={(e) => setBody(e.target.value)}
        placeholder="write post body..."
        rows={5}
        data-testid="body-input"
      />
      <button
        className={classes.button}
        type="submit"
        disabled={!body.length || !title.length}
        data-testid="submit-button"
      >
        {isLoading ? "Submitting..." : id ? "Edit" : "Create"}
      </button>
    </form>
  );
};

export default PostForm;
