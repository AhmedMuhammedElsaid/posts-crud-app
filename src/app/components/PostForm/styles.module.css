.form {
  display: flex;
  flex-direction: column;
  width: 350px;
  margin: 20px auto;
  gap: 10px;
}

.input {
  margin-bottom: 20px;
  padding: 10px;
  font-size: 16px;
  border-radius: 20px;
}

.textarea {
  margin-bottom: 20px;
  padding: 10px;
  font-size: 16px;
  border-radius: 20px;
  resize: none;
}
.label{
    font-size: 16px;
    color: #342c2cd1;
    font-weight: bold;
    margin-inline-start: 5px;
}
.button {
  padding: 10px;
  font-size: 16px;
  background-color: #0070f3;
  color: white;
  border: none;
  border-radius: 20px;
  cursor: pointer;
  height: 45px;
}
.button:disabled {
  opacity: 0.5;
  pointer-events: none;
}
.header {
  font-size: 26px;
  color: #333333c7;
  padding: 10px;
  text-align: center;
}
