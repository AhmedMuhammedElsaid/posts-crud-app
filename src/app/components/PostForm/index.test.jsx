import React from "react";
import { render, fireEvent, waitFor, screen } from "@testing-library/react";
import axios from "axios"; // You might want to consider using a testing library for mocking HTTP requests, like `axios-mock-adapter`.
import { toast } from "react-toastify";
import PostForm from "./index.tsx";

// Mocking axios to prevent actual HTTP requests during testing
jest.mock("axios");

describe("test PostForm component scenaries", () => {
  beforeEach(() => {
    // Mocking the toast methods to prevent actual toast notifications during testing
    jest.spyOn(toast, "success").mockImplementation(() => {});
    jest.spyOn(toast, "error").mockImplementation(() => {});
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  test("renders PostForm with Create button and submits form", async () => {
    // Mocking a successful POST request
    axios.mockResolvedValueOnce({ status: 201 });

    const { asFragment } = render(<PostForm />);
    expect(asFragment()).toMatchSnapshot();
    const titleInput = screen.getByTestId("title-input");
    const bodyInput = screen.getByTestId("body-input");
    const submitButton = screen.getByTestId("submit-button");

    // Filling in form inputs
    fireEvent.change(titleInput, { target: { value: "Test Title" } });
    fireEvent.change(bodyInput, { target: { value: "Test Body" } });

    // Triggering form submission
    fireEvent.click(submitButton);

    // Waiting for the asynchronous actions to complete
    await waitFor(() => {
      expect(axios).toHaveBeenCalledWith(expect.stringContaining("posts"), {
        method: "POST",
        data: { title: "Test Title", body: "Test Body", userId: 1 },
      });
      // eslint-disable-next-line testing-library/no-wait-for-multiple-assertions
      expect(toast.success).toHaveBeenCalledWith("Post Created Successfully");
    });
  });
  test("renders PostForm with Edit button and submits form for editing", async () => {
    // Mocking a successful PUT request
    // axios.mockResolvedValueOnce({ status: 201 });
    // Mocking a successful GET request with specific post data
    const mockPostData = {
      data: {
        id: 1,
        title: "Test Title123",
        body: "Test Body123",
        userId: 1,
      },
      status: 200,
    };

    axios.mockResolvedValueOnce(mockPostData);

    render(<PostForm id={'1'} />);
  
    const titleInput = screen.getByTestId("title-input");
    const bodyInput = screen.getByTestId("body-input");
    const submitButton = screen.getByTestId("submit-button");

    // Filling in form inputs
    fireEvent.change(titleInput, { target: { value: "Test Title" } });
    fireEvent.change(bodyInput, { target: { value: "Test Body" } });

    // Triggering form submission
    fireEvent.click(submitButton);

    // Waiting for the asynchronous actions to complete
    await waitFor(() => {
      expect(axios).toHaveBeenCalledWith(expect.stringContaining("https://jsonplaceholder.typicode.com/posts"), {
        method: "PUT",
        data: { title: "Test Title", body: "Test Body", userId: 1,id:"1" },
      });
      // eslint-disable-next-line testing-library/no-wait-for-multiple-assertions
      expect(toast.success).toHaveBeenCalledWith(`Post Number 1 Edited Successfully`);
    });
    
  });
});
