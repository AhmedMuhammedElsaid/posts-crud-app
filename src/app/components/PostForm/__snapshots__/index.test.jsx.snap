// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test PostForm component scenaries renders PostForm with Create button and submits form 1`] = `
<DocumentFragment>
  <form
    class="form"
  >
    <label
      class="label"
      for="body"
    >
      Title
    </label>
    <input
      class="input"
      data-testid="title-input"
      placeholder="write post title..."
      type="text"
      value=""
    />
    <label
      class="label"
      for="body"
    >
      Body
    </label>
    <textarea
      class="textarea"
      data-testid="body-input"
      placeholder="write post body..."
      rows="5"
    />
    <button
      class="button"
      data-testid="submit-button"
      disabled=""
      type="submit"
    >
      Create
    </button>
  </form>
</DocumentFragment>
`;
